infer/environment.cc

```diff
@@ -513,11 +556,13 @@ void Environment::updateKnowledge(core::Context ctx, cfg::LocalRef local, core::
         return;
     }
 
+    if (!knowledgeFilter.isNeeded(local)) {
+        return;
+    }
```

Comment:
Have you checked what kind of effect this has on typechecking performance?

I worry about this being extra work...but thinking about it a little harder, since this code is guarded by `isUpdateKnowledgeName`, presumably we were doing this work anyway, and this just commonizes the code.